fix(core): make injector.get() return default value with InjectFlags.…

thekiba opened the pull request:
…Self flag on

Fixes #27729

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #27729 


## What is the new behavior?

injector.get() return default value with InjectFlags.Self flag on

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
**injector.get()** will return default value with **InjectFlags.Self** flag on instead of **undefined**

## Other information

JoostK commented on packages/core/src/di/injector.ts:
```diff
@@ -344,6 +344,8 @@ function resolveToken(
     }
   } else if (!(flags & InjectFlags.Self)) {
     value = parent.get(token, notFoundValue, InjectFlags.Default);
+  } else if (flags & InjectFlags.Self) {
```
Redundant guard:

```suggestion
   } else {
```

Perhaps it would be better to swap this if with the preceding one.

thekiba replied:
I like it! I think all this will soon be deprecated, so I made it into a separate condition with a specific problem.

Also it can be assigns a default value during initialization. What do you think about that?
```typescript
let value = notFoundValue;
```

JoostK replied:
I prefer the explicitness of the additional `else` statement.

thekiba commented:
I signed it!

thekiba commented:
@pkozlowski-opensource I don't know what googlebot wants. It is waiting for confirmation or isn't?

JoostK commented:
@thekiba googlebot can't verify the CLA as this PR now contains commits from various authors. This is typically not a problem and will be resolved by a Googler (by hand).

Anyway, the second commit that contains my suggestion can be folded into the initial commit. We'll want for this change to only be a single commit anyway.

Locally, please pull the changes from this branch and get creative with Git. You can rebase, but the easiest way for me to explain here would be to softly reset to your commit (`HEAD~`) and amend my changes (then in the index because of the soft reset) to the first commit:

```
git pull
git reset --soft HEAD~
git commit --amend
git push --force-with-lease
```

That should do it.

mhevery commented on packages/core/test/di/static_injector_spec.ts:
```diff
@@ -418,6 +418,11 @@ function factoryFn(a: any){}
+
+      it('should return a default value when not requested provider on self', () => {
+        const injector = Injector.create([]);
+        expect(injector.get<Car|null>(Car, null, InjectFlags.Self)).toBeNull();
```
The above implementation is not quite right If you add additional tests.
```
expect(() => injector.get<Car|null>(Car, undefined, InjectFlags.Self)).toThrow('....');
```
You will see that it will not pass. Could you fix it?

thekiba replied:
Did it! 🙌

Also I added some more tests. For example, should return null when there is no default value and the Self and Optional flag on.

mhevery commented on packages/core/src/di/injector.ts:
```diff
@@ -336,6 +336,8 @@ function resolveToken(
   } else if (!(flags & InjectFlags.Self)) {
     value = parent.get(token, notFoundValue, InjectFlags.Default);
+  } else {
+    value = notFoundValue;
```
I think this should delegate to `NullInjector.get` so that there is a chance to throw a proper error.

thekiba replied:
I delegate it to NullInjector and added additional check for Optional flag

mhevery commented:
Also pleases rebase on the latest master and squash commits into single commit.

thekiba committed:
fix(core): make injector.get() return default value with InjectFlags.Self flag on

Fixes #27729

mhevery commented:
http://test/OCL:260227161:BASE:260227188:1564180732811:4d324ea9

AndrewKushnir commented:
@thekiba thanks for the fix!

It looks like the tests are failing because `NULL_INJECTOR` was not imported. Could you please have a look?

Thank you.

thekiba commented:
@AndrewKushnir I have replaced `NULL_INJECTOR` to `Injector.NULL` and fixed it.

@mhevery @AndrewKushnir I found a bug with Ivy R3Injector, it doesn't return a default value when not requested provider on self and optional flags. I fixed it too.

And I found another one with Ivy. `R3Injector.get()`, `NgModuleRef.get()`, etc, sets `notFoundValue` as `THROW_IF_NOT_FOUND` when it `undefined`. In this case, we cannot return the expected `undefined` value, because it sets as `THROW_IF_NOT_FOUND`.

AndrewKushnir commented:
@thekiba thanks for additional fixes. It looks like `test_ivy_aot` is failing with injector-related error, which might potentially be related to your changes. Could you please have a look?

We may also need to ask @mhevery to have a quick look at the changes once again when the mentioned failure is fixes, since you added some extra code to `r3_injector` as you pointed out.

Thank you.

thekiba commented:
@AndrewKushnir thanks! `test_ivy_aot` is failing because I added extra test for `static_injector_spec.ts` witch successfully passed for not Ivy injector. I added a fix for it. `notFoundValue` checks as `THROW_IF_NOT_FOUND` instead of `undefined`.

Also, I have a few problems with Bazel and gulp locally. Can we discuss that somewhere?

AndrewKushnir commented:
@mhevery could you please have another look at the changes in this PR? Thank you.

AndrewKushnir commented:
@thekiba thanks for additional fixes!

I think you can open a ticket using "Issues" section and describe the problem there (with a repro if possible), so the team and community members can have a look and try to help.

Thank you.

thekiba commented:
I'm still having this pr

AndrewKushnir commented:
Update:

- Blueprint and Global TAP Presubmits for both VE and Ivy are successful
- CI is green
- Misko approved this PR
- I've changed the label to "master-only" (so this change would be a part of v9 release), since this is still a slight behavior change

Based on this, I'm adding "merge" label back.

Thank you.
